File: client/js/character.js

define(['entity', 
        'transition', 
        'timer'], function(
        Entity, 
        Transition, 
        Timer) {

    var Character = Entity.extend({
        data: {
            // Health
            hp: 0,
            maxHP: 0,
           
            // Level
            level: 1,

            // Gear
            weapon: null,
            armor: null
        },

        init: function(id, kind) {
    	    var self = this;
            this._super(id, kind);
        
            this._hp = 0;
            this._maxHP = 0;
            this._level = 1;
            this._weapon = null;
            this._armor = null;

    		// Position and orientation
    		this.nextGridX = -1;
    		this.nextGridY = -1;
    		this.orientation = Types.Orientations.DOWN;
		
    		// Speeds
            this.atkSpeed = 50;
    		this.moveSpeed = 120;
    		this.walkSpeed = 100;
    		this.idleSpeed = 450;
    		this.setAttackRate(800);
        
            // Pathing
    		this.movement = new Transition();
    		this.path = null;
    		this.newDestination = null;
    		this.adjacentTiles = {};
		
    		// Combat
    		this.target = null;
            this.unconfirmedTarget = null;
            this.attackers = {};
       
            // Modes
            this.isDead = false;
            this.attackingMode = false;
            this.followingMode = false;
    	},
	
    	clean: function() {
    	    this.forEachAttacker(function(attacker) {
                attacker.disengage();
                attacker.idle();
            });
    	},

        get hp() {
            return this._hp;
        },

        set hp(hp) {
            this._hp = hp;
            this.trigger("change");
        },

        get maxHP() {
            return this._maxHP;
        },

        set maxHP(maxHP) {
            this._maxHP = maxHP;
            this.trigger("change");
        },
        
        get level() {
            return this._level;
        },

        set level(level) {
            this._level = level;
            this.trigger("change");
        },
   
        get armor() {
            return this._armor;
        },
 
        get skin() {
            if (!this.armor) {
                return this.kind;
            }

            return this.armor;
        },

        set armor(armor) {
            this._armor = armor;
            this.trigger("change");
        },

        get weapon() {
            return this._weapon;
        },

        set weapon(weapon) {
            this._weapon = weapon;
            this.trigger("change");
        },

        equipArmor: function(kind) {
            this.armor = kind;
        },

        getArmorLevel: function() {
            return Properties.getArmorLevel(this.armor);
        },
      
        equipWeapon: function(kind) {
            this.weapon = kind;
            this.setWeaponName(Types.getKindAsString(this.weapon));
        },

        getWeaponLevel: function() {
            return Properties.getWeaponLevel(this.weapon);
        },
        
    	setDefaultAnimation: function() {
    		this.idle();
    	},
	
    	hasWeapon: function() {
    	    return false;
    	},
	
    	hasShadow: function() {
    	    return true;
    	},
	
    	animate: function(animation, speed, count, onEndCount) {
    	    var oriented = ['atk', 'walk', 'idle'];
    	        o = this.orientation;
            
            if(!(this.currentAnimation && this.currentAnimation.name === "death")) { // don't change animation if the character is dying
            	this.flipSpriteX = false;
        	    this.flipSpriteY = false;
	    
        	    if(_.indexOf(oriented, animation) >= 0) {
        	        animation += "_" + (o === Types.Orientations.LEFT ? "right" : Types.getOrientationAsString(o));
        	        this.flipSpriteX = (this.orientation === Types.Orientations.LEFT) ? true : false;
        	    }

        		this.setAnimation(animation, speed, count, onEndCount);
        	}
    	},
    
        turnTo: function(orientation) {
            this.orientation = orientation;
            this.idle();
        },
	
    	setOrientation: function(orientation) {
    	    if(orientation) {
    	        this.orientation = orientation;
    	    }
    	},
	
    	idle: function(orientation) {
    	    this.setOrientation(orientation);
    	    this.animate("idle", this.idleSpeed);
    	},
	
    	hit: function(orientation) {
    	    this.setOrientation(orientation);
    	    this.animate("atk", this.atkSpeed, 1);
    	},
	
    	walk: function(orientation) {
    	    this.setOrientation(orientation);
    	    this.animate("walk", this.walkSpeed);
    	},
    
        moveTo_: function(x, y, callback) {
            this.destination = { gridX: x, gridY: y };
            this.adjacentTiles = {};
        
            if (this.isMoving()) {
                this.continueTo(x, y);
            }
            else {
                var path = this.requestPathfindingTo(x, y);
            
                this.followPath(path);
            }
        },
    
        requestPathfindingTo: function(x, y) {
            var ignored = [this]; // Always ignore self
            var target = null;
            if (this.hasTarget()) {
                target = this.target; 
            } else if(this.previousTarget) {
                // If repositioning before attacking again, ignore previous target
                // See: tryMovingToADifferentTile()
                target = this.previousTarget;
            }
            if (target) {
                ignored.push(target);

                // also ignore other attackers of the target entity
                target.forEachAttacker(function(attacker) {
                    ignored.push(attacker);
                });
            }
            
            return globalGame.findPath(this, x, y, ignored);
        },

        startPathing: function(path) {

        },
        stopPathing: function(x, y) {
            if (!this.isDying) {
                if (this.hasTarget() && this.isAdjacent(this.target)) {
                    this.lookAtTarget();
                }

                var self = this;
                this.forEachAttacker(function(attacker) {
                    if (!attacker.isAdjacentNonDiagonal(self) && attacker.id !== self.id) {
                        attacker.follow(self);
                    }
                });
    
                globalGame.unregisterEntityPosition(this);
                globalGame.registerEntityPosition(this);
            }
        },
    
    	followPath: function(path) {
    		if(path.length > 1) { // Length of 1 means the player has clicked on himself
    			this.path = path;
    			this.step = 0;
			
                if(this.followingMode) { // following a character
                    path.pop();
                }
			
                this.startPathing(path);
                this.nextStep();
    		}
    	},

    	continueTo: function(x, y) {
    		this.newDestination = { x: x, y: y };
    	},
	
    	updateMovement: function() {
    		var p = this.path,
    			i = this.step;
		
    		if(p[i][0] < p[i-1][0]) {
    			this.walk(Types.Orientations.LEFT);
    		}
    		if(p[i][0] > p[i-1][0]) {
    			this.walk(Types.Orientations.RIGHT);
    		}
    		if(p[i][1] < p[i-1][1]) {
    			this.walk(Types.Orientations.UP);
    		}
    		if(p[i][1] > p[i-1][1]) {
    			this.walk(Types.Orientations.DOWN);
    		}
    	},

    	updatePositionOnGrid: function() {
    		this.setGridPosition(this.path[this.step][0], this.path[this.step][1]);
    	},

    	nextStep: function() {
            var stop = false,
                x, y, path;
        
    		if(this.isMoving()) {
                this.beforeStep();
            
    		    this.updatePositionOnGrid();
                this.checkAggro();
            
                if(this.interrupted) { // if Character.stop() has been called
                    stop = true;
                    this.interrupted = false;
                }
                else {
                    if(this.hasNextStep()) {
                        this.nextGridX = this.path[this.step+1][0];
                        this.nextGridY = this.path[this.step+1][1];
                    }
            
                    this.doStep();
			    
        			if(this.hasChangedItsPath()) {
        				x = this.newDestination.x;
        				y = this.newDestination.y;
        				path = this.requestPathfindingTo(x, y);
                
                        this.newDestination = null;
        				if(path.length < 2) {
                            stop = true;
                        }
                        else {
                            this.followPath(path);
                        }
        			}
        			else if(this.hasNextStep()) {
        				this.step += 1;
        				this.updateMovement();
        			}
                    else {
                        stop = true;
                    }
                }
            
    		    if(stop) { // Path is complete or has been interrupted
    		        this.path = null;
        			this.idle();
                
                    this.stopPathing(this.gridX, this.gridY);
        		}
        	}
    	},
    
        beforeStep: function() {
            globalGame.unregisterEntityPosition(this);
        },

        doStep: function() {
            if (!this.isDying) {
                globalGame.registerEntityDualPosition(this);

                var self = this;
                this.forEachAttacker(function(attacker) {
                    if (attacker.isAdjacent(attacker.target)) {
                        attacker.lookAtTarget();
                    } else {
                        attacker.follow(self);
                    }
                });
            }
        },

    	isMoving: function() {
    		return !(this.path === null);
    	},

    	hasNextStep: function() {
    		return (this.path.length - 1 > this.step);
    	},

    	hasChangedItsPath: function() {
    		return !(this.newDestination === null);
        },
    
        isNear: function(character, distance) {
            var dx, dy, near = false;
        
            dx = Math.abs(this.gridX - character.gridX);
            dy = Math.abs(this.gridY - character.gridY);
        
            if(dx <= distance && dy <= distance) {
                near = true;
            }
            return near;
        },
    
        checkAggro: function() {
        },
        
        aggro: function(character) {
        },
    
        /**
         * Changes the character's orientation so that it is facing its target.
         */
        lookAtTarget: function() {
            if(this.target) {
                this.turnTo(this.getOrientationTo(this.target));
            }
        },
    
        /**
         * 
         */
        go: function(x, y) {
    	    if(this.isAttacking()) {
                this.disengage();
            }
            else if(this.followingMode) {
                this.followingMode = false;
                this.target = null;
            }
            this.moveTo_(x, y);
        },
    
        /**
         * Makes the character follow another one.
         */
        follow: function(entity) {
            if(entity) {
                this.followingMode = true;
                this.moveTo_(entity.gridX, entity.gridY);
            }
        },
    
        /**
    	 * Stops a moving character.
    	 */
    	stop: function() {
    	    if(this.isMoving()) {
                this.interrupted = true;
    	    }
    	},
    
        /**
         * Makes the character attack another character. Same as Character.follow but with an auto-attacking behavior.
         * @see Character.follow
         */
        engage: function(character) {
            this.attackingMode = true;
            this.setTarget(character);
            this.follow(character);
        },
    
        disengage: function() {
            this.attackingMode = false;
            this.followingMode = false;
            this.removeTarget();
        },
    
        /**
         * Returns true if the character is currently attacking.
         */
        isAttacking: function() {
            return this.attackingMode;
        },
        
        /**
         * Gets the right orientation to face a target character from the current position.
         * Note:
         * In order to work properly, this method should be used in the following
         * situation :
         *    S
         *  S T S
         *    S
         * (where S is self, T is target character)
         * 
         * @param {Character} character The character to face.
         * @returns {String} The orientation.
         */
        getOrientationTo: function(character) {
            if(this.gridX < character.gridX) {
                return Types.Orientations.RIGHT;
            } else if(this.gridX > character.gridX) {
                return Types.Orientations.LEFT;
            } else if(this.gridY > character.gridY) {
                return Types.Orientations.UP;
            } else {
                return Types.Orientations.DOWN;
            }
        },
    
        /**
         * Returns true if this character is currently attacked by a given character.
         * @param {Character} character The attacking character.
         * @returns {Boolean} Whether this is an attacker of this character.
         */
        isAttackedBy: function(character) {
            return (character.id in this.attackers);
        },
    
        /**
        * Registers a character as a current attacker of this one.
        * @param {Character} character The attacking character.
        */
        addAttacker: function(character) {
            if(!this.isAttackedBy(character)) {
                this.attackers[character.id] = character;
            } else {
                log.error(this.id + " is already attacked by " + character.id);
            }
        },
    
        /**
        * Unregisters a character as a current attacker of this one.
        * @param {Character} character The attacking character.
        */
        removeAttacker: function(character) {
            if(this.isAttackedBy(character)) {
                delete this.attackers[character.id];
            } else {
       //         log.error(this.id + " is not attacked by " + character.id);
            }
        },
    
        /**
         * Loops through all the characters currently attacking this one.
         * @param {Function} callback Function which must accept one character argument.
         */
        forEachAttacker: function(callback) {
            _.each(this.attackers, function(attacker) {
                callback(attacker);
            });
        },
    
        /**
         * Sets this character's attack target. It can only have one target at any time.
         * @param {Character} character The target character.
         */
        setTarget: function(character) {
            if(this.target !== character) { // If it's not already set as the target
                if(this.hasTarget()) {
                    this.removeTarget(); // Cleanly remove the previous one
                }
                this.unconfirmedTarget = null;
                this.target = character;
            } else {
                log.debug(character.id + " is already the target of " + this.id);
            }
        },
    
        /**
         * Removes the current attack target.
         */
        removeTarget: function() {
            var self = this;
        
            if(this.target) {
                if(this.target instanceof Character) {
                    this.target.removeAttacker(this);
                }
                this.target = null;
            }
        },
    
        /**
         * Returns true if this character has a current attack target.
         * @returns {Boolean} Whether this character has a target.
         */
        hasTarget: function() {
            return !(this.target === null);
        },

        /**
         * Marks this character as waiting to attack a target.
         * By sending an "attack" message, the server will later confirm (or not)
         * that this character is allowed to acquire this target.
         *
         * @param {Character} character The target character
         */
        waitToAttack: function(character) {
            this.unconfirmedTarget = character;
        },
    
        /**
         * Returns true if this character is currently waiting to attack the target character.
         * @param {Character} character The target character.
         * @returns {Boolean} Whether this character is waiting to attack.
         */
        isWaitingToAttack: function(character) {
            return (this.unconfirmedTarget === character);
        },
    
        /**
         * 
         */
        canAttack: function(time) {
            if(this.canReachTarget() && this.attackCooldown.isOver(time)) {
                return true;
            }
            return false;
        },
        
        canReachTarget: function() {
            if(this.hasTarget() && this.isAdjacentNonDiagonal(this.target)) {
                return true;
            }
            return false;
        },
    
        /**
         * 
         */
    	die: function() {
    	    this.removeTarget();
    	    this.isDead = true;

            log.info(this.id + " is dead");
        
            this.isDying = true;
            this.setSprite(globalGame.sprites["death"]);

            var self = this;
            this.animate("death", 120, 1, function() {
                log.info(self.id + " was removed");

                globalGame.removeEntity(self);
                globalGame.removeFromRenderingGrid(self, self.gridX, self.gridY);
            });

            this.forEachAttacker(function(attacker) {
                attacker.disengage();
            });
            
            if (globalGame.player.target 
               && globalGame.player.target.id === this.id) 
            {
                globalGame.player.disengage();
            }
        
            // Upon death, this entity is removed from both grids, allowing the player
            // to click very fast in order to loot the dropped item and not be blocked.
            // The entity is completely removed only after the death animation has ended.
            globalGame.removeFromEntityGrid(this, this.gridX, this.gridY);
            globalGame.removeFromPathingGrid(this.gridX, this.gridY);
        
            if (globalGame.camera.isVisible(this)) {
                globalGame.audioManager.playSound("kill"+Math.floor(Math.random()*2+1));
            }
        
            globalGame.updateCursor();
    	},

        moved: function() {
            this.dirty();
            // Make chat bubbles follow moving entities
            globalGame.assignBubbleTo(this); 
        },
	
    	hurt: function() {
            var self = this;
        
            this.stopHurting();
            this.sprite = this.hurtSprite;
            this.hurting = setTimeout(this.stopHurting.bind(this), 75);
        },
    
        stopHurting: function() {
            this.sprite = this.normalSprite;
            clearTimeout(this.hurting);
        },
    
        setAttackRate: function(rate) {
            this.attackCooldown = new Timer(rate);
        }
    });
    
    return Character;
});

File: client/dart/character.dart
library character;

import "dart:async";
import "dart:math";

import "entity.dart";
import "game.dart";
import "position.dart";
import "sprite.dart";
import "transition.dart";
import "lib/gametypes.dart";

class Character extends Entity {

  int _hp = 0;
  int _maxHP = 0;
  int _level = 0;

  EntityKind _weapon;
  EntityKind _armor;

  // Position and orientation
  int nextGridX = -1;
  int nextGridY = -1;

  // Speeds
  int atkSpeed = 50;
  int moveSpeed = 120;
  int walkSpeed = 100;
  int idleSpeed = 450;
  int attackRate = 800;
  Timer attackCooldown;

  // Pathing
  Transition movement = new Transition();
  List<List<int>> path;
  int step = 0;
  Position destination;
  Position newDestination;

  // Combat
  Character _target;
  Character previousTarget;
  Map<int, Character> attackers = {};

  // Modes
  bool isMovementInterrupted = false;
  bool attackingMode = false;
  bool followingMode = false;
  bool showWeapon = true;
  bool showArmor = true;

  bool isHurting = false;
  Timer hurtingTimer;

  Character(int id, EntityKind kind): super(id, kind);

  void reset() {
    super.reset();

    // Position and orientation
    this.nextGridX = -1;
    this.nextGridY = -1;

    // Speeds
    this.atkSpeed = 50;
    this.moveSpeed = 120;
    this.walkSpeed = 100;
    this.idleSpeed = 450;
    this.attackRate = 800;

    // Pathing
    this.movement = new Transition();
    this.path = null;
    this.newDestination = null;

    // Combat
    this.target = null;
    this.attackers = new Map<int, Entity>();

    // Modes
    this.attackingMode = false;
    this.followingMode = false;
  }

  void clean() {
    super.clean();

    this.forEachAttacker((Character attacker) {
      attacker.disengage();
      attacker.idle();
    });
  }

  Sprite get sprite {
    Sprite sprite = super.sprite;
    if (this.armor != null) {
      String kindString = Types.getKindAsString(this.armor);
      sprite = Game.sprites[kindString];
    }

    return this.isHurting ? sprite.getHurtSprite() : sprite;
  }

  bool isHostile(Entity entity) => false;

  Character get target => this._target;
  void set target(Character target) {
    this._target = target;
    this.trigger("TargetChange");
    this.trigger("change");
  }

  int get hp => this._hp;
  void set hp(int hp) {
    this._hp = hp;
    this.trigger("HealthChange");
    this.trigger("change");
  }

  int get maxHP => this._maxHP;
  void set maxHP(int maxHP) {
    this._maxHP = maxHP;
    this.trigger("HealthChange");
    this.trigger("change");
  }

  int get hpPercentage => this.maxHP == 0 ? 0 : (this.hp * 100 / this.maxHP).floor();

  int get level => this._level;
  void set level(int level) {
    this._level = level;
    this.trigger("LevelChange");
    this.trigger("change");
  }

  EntityKind get armor => this._armor;
  void set armor(EntityKind armor) {
    this._armor = armor;
    this.trigger("ArmorChange");
    this.trigger("EquipmentChange");
    this.trigger("change");
  }

  EntityKind get weapon => this._weapon;
  void set weapon(EntityKind weapon) {
    this._weapon = weapon;
    this.trigger("WeaponChange");
    this.trigger("EquipmentChange");
    this.trigger("change");
  }

  int getArmorRank() => Types.getArmorRank(this.armor);

  int getWeaponRank() => Types.getWeaponRank(this.weapon);

  bool hasWeapon() => this.weapon != null;

  void setDefaultAnimation() {
    this.idle();
  }

  bool hasShadow() => false;

  void turnTo(Orientation orientation) {
    this.orientation = orientation;
    this.idle();
  }

  void idle([Orientation orientation]) {
    if (orientation != null) {
      this.orientation = orientation;
    }

    this.animate("idle", this.idleSpeed);
  }

  void hit([Orientation orientation]) {
    if (orientation != null) {
      this.orientation = orientation;
    }

    this.attackCooldown = new Timer(new Duration(milliseconds: this.attackRate), () {});
    this.animate("atk", this.atkSpeed, 1);
  }

  void walk([Orientation orientation]) {
    if (orientation != null) {
      this.orientation = orientation;
    }

    this.animate("walk", this.walkSpeed);
  }

  void moveTo_(Position position) {
    this.destination = position;

    if (this.isMoving()) {
      this.continueTo(position);
      return;
    }

    this.followPath(this.requestPathfindingTo(position));
  }

  List<List<int>> requestPathfindingTo(Position position) {
    return Game.findPath(this, position, new List<Character>()..add(this));
  }

  void startPathing(List<List<int>> path) {
  }

  void stopPathing(Position position) {
    if (this.isDying) {
      return;
    }

    if (this.hasTarget() && this.isAdjacent(this.target)) {
      this.lookAtTarget();
    }

    this.forEachAttacker((Character attacker) {
      if (!attacker.isAdjacentNonDiagonal(this) && attacker.id != this.id) {
        attacker.follow(this);
      }
    });

    Game.unregisterEntityPosition(this);
    Game.registerEntityPosition(this);
  }

  void followPath(List path) {
    if (path.length <= 1) { // Length of 1 means the player has clicked on himself
      return;
    }

    this.path = path;
    this.step = 0;

    if (this.followingMode) { // following a character
      path.removeLast();
    }

    this.startPathing(path);
    this.nextStep();
  }

  void continueTo(Position position) {
    this.newDestination = position;
  }

  void updateMovement() {
    if (this.path[this.step][0] < this.path[this.step - 1][0]) {
      this.walk(Orientation.LEFT);
    }
    if (this.path[this.step][0] > this.path[this.step - 1][0]) {
      this.walk(Orientation.RIGHT);
    }
    if (this.path[this.step][1] < this.path[this.step - 1][1]) {
      this.walk(Orientation.UP);
    }
    if (this.path[this.step][1] > this.path[this.step - 1][1]) {
      this.walk(Orientation.DOWN);
    }
  }

  void updatePositionOnGrid() {
    this.gridPosition = new Position(this.path[this.step][0], this.path[this.step][1]);
  }

  void nextStep() {
    bool stop = false;
    List<List<int>> path;

    if (!this.isMoving()) {
      return;
    }

    this.beforeStep();

    this.updatePositionOnGrid();

    if (this.isMovementInterrupted) { // if Character.stop() has been called
      stop = true;
      this.isMovementInterrupted = false;
    } else {
      if (this.hasNextStep()) {
        this.nextGridX = this.path[this.step + 1][0];
        this.nextGridY = this.path[this.step + 1][1];
      }

      this.doStep();

      if (this.hasChangedItsPath()) {
        path = this.requestPathfindingTo(this.newDestination);

        this.newDestination = null;
        if (path.length < 2) {
          stop = true;
        } else {
          this.followPath(path);
        }
      } else if (this.hasNextStep()) {
        this.step++;
        this.updateMovement();
      } else {
        stop = true;
      }
    }

    if (stop) { // Path is complete or has been interrupted
      this.path = null;
      this.idle();

      this.stopPathing(this.gridPosition);
    }
  }

  void beforeStep() {
    Game.unregisterEntityPosition(this);
  }

  void doStep() {
    if (this.isDying) {
      return;
    }

    Game.registerEntityDualPosition(this);

    this.forEachAttacker((attacker) {
      if (attacker.isAdjacent(attacker.target)) {
        attacker.lookAtTarget();
      } else {
        attacker.follow(this);
      }
    });
  }

  bool isMoving() => (this.path != null);

  bool hasNextStep() => (this.path.length - 1 > this.step);

  bool hasChangedItsPath() => (this.newDestination != null);

  bool isNear(Character character, num distance) =>
    (this.gridPosition.x - character.gridPosition.x).abs() <= distance
    && (this.gridPosition.y - character.gridPosition.y).abs() <= distance;

  void lookAtTarget() {
    if (this.hasTarget()) {
      this.turnTo(this.getOrientationTo(this.target));
    }
  }

  void go(Position position) {
    if (this.isAttacking()) {
      this.disengage();
    } else if (this.followingMode) {
      this.followingMode = false;
      this.target = null;
    }

    this.moveTo_(position);
  }

  void follow(Entity entity) {
    this.followingMode = true;
    this.moveTo_(entity.gridPosition);
  }

  void stop() {
    if (this.isMoving()) {
      this.isMovementInterrupted = true;
    }
  }

  /**
   * Makes the character attack another character. Same as Character.follow but with an auto-attacking behavior.
   * @see Character.follow
   */
  void engage(Character character) {
    this.attackingMode = true;
    this.setTarget(character);
    this.follow(character);
  }

  void disengage() {
    this.attackingMode = false;
    this.followingMode = false;
    this.removeTarget();
  }

  bool isAttacking() => this.attackingMode;

  /**
   * Gets the right orientation to face a target character from the current position.
   * Note:
   * In order to work properly, this method should be used in the following
   * situation :
   *    S
   *  S T S
   *    S
   * (where S is self, T is target character)
   *
   * @param {Character} character The character to face.
   * @returns {String} The orientation.
   */
  Orientation getOrientationTo(Character character) {
    if (this.gridPosition.x < character.gridPosition.x) {
      return Orientation.RIGHT;
    } else if (this.gridPosition.x > character.gridPosition.x) {
      return Orientation.LEFT;
    } else if (this.gridPosition.y > character.gridPosition.y) {
      return Orientation.UP;
    }

    return Orientation.DOWN;
  }

  bool isAttackedBy(character) => this.attackers.containsKey(character.id);

  void addAttacker(Character character) {
    this.attackers.putIfAbsent(character.id, () => character);
  }

  void removeAttacker(Character character) {
    this.attackers.remove(character.id);
  }

  /**
   * Loops through all the characters currently attacking this one.
   * @param {Function} callback Function which must accept one character argument.
   */
  void forEachAttacker(void callback(Character character)) {
    Map<int, Character> attackers = new Map.from(this.attackers);
    attackers.forEach((int id, Character character) {
      callback(character);
    });
  }

  void setTarget(Entity entity) {
    if (this.hasTarget() && this.target == entity) {
      // If it's not already set as the target
      return;
    }

    if (this.hasTarget()) {
      this.removeTarget(); // Cleanly remove the previous one
    }
    this.target = entity;
  }

  void removeTarget() {
    if (!this.hasTarget()) {
      return;
    }

    this.target.removeAttacker(this);
    this.target = null;
  }

  bool hasTarget() => (this.target != null);

  bool canAttack(int time) =>
      this.canReachTarget()
      && (this.attackCooldown == null || !this.attackCooldown.isActive);

  bool canReachTarget() => (this.hasTarget() && this.isAdjacentNonDiagonal(this.target));

  void die() {
    super.die();
    this.removeTarget();

    this.forEachAttacker((Character attacker) {
      attacker.disengage();
    });

    if (Game.player.target != null && Game.player.target.id == this.id) {
      Game.player.disengage();
    }

    if (Game.camera.isVisible(this)) {
      Random rng = new Random();
      Game.audioManager.playSound("kill${rng.nextInt(2)+1}");
    }
  }

  void moved() {
    this.dirty();

    // Make chat bubbles follow moving entities
    Game.assignBubbleTo(this);
  }

  void hurt() {
    this.stopHurting();
    this.isHurting = true;
    this.hurtingTimer = new Timer(new Duration(milliseconds: 75), this.stopHurting);
  }

  void stopHurting() {
    this.isHurting = false;
    if (this.hurtingTimer != null && this.hurtingTimer.isActive) {
      this.hurtingTimer.cancel();
    }
  }
}
